core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java

```diff
@@ -159,18 +169,124 @@ public void commit(TableMetadata base, TableMetadata metadata) {
+    int previousVersionsMax =
+        metadata.propertyAsInt(
+            TableProperties.METADATA_PREVIOUS_VERSIONS_MAX,
+            TableProperties.METADATA_PREVIOUS_VERSIONS_MAX_DEFAULT);
+    // todo:Currently, if the user is using an object store, we assume that he must be using the
+    //  global locking service. But we should support add some other conditions in future.
+    boolean supportGlobalLocking = useObjectStore;
+    try {
+      tryLock(tempMetadataFile, metadataRoot());
+      versionCommitSuccess =
+          commitNewVersion(
+              fs, tempMetadataFile, finalMetadataFile, nextVersion, supportGlobalLocking);
+      if (!versionCommitSuccess) {
+        throw new CommitFailedException(
+            "Can not commit newMetaData because version [%s] has already been committed. tempMetaData=[%s],finalMetaData=[%s].Are there other clients running in parallel with the current task?",
+            nextVersion, tempMetadataFile, finalMetadataFile);
+      }
+      validate(supportGlobalLocking, previousVersionsMax, nextVersion, fs, finalMetadataFile);
+      this.shouldRefresh = true;
+      LOG.info("Committed a new metadata file {}", finalMetadataFile);
+      // update the best-effort version pointer
+      writeVersionHint(fs, nextVersion);
+      deleteRemovedMetadataFiles(base, metadata);
+    } catch (CommitStateUnknownException e) {
+      this.shouldRefresh = true;
+      throw e;
+    } catch (Exception e) {
+      this.shouldRefresh = versionCommitSuccess;
+      if (!versionCommitSuccess) {
+        tryDelete(tempMetadataFile);
+        throw new CommitFailedException(e);
+      }
+    } finally {
+      unlock(tempMetadataFile, metadataRoot());
+    }
+  }
 
-    // this rename operation is the atomic commit operation
-    renameToFinal(fs, tempMetadataFile, finalMetadataFile, nextVersion);
+  private void tryDelete(Path path) {
+    try {
+      io().deleteFile(path.toString());
+    } catch (Exception ignored) {
+      // do nothing
+    }
+  }
 
-    LOG.info("Committed a new metadata file {}", finalMetadataFile);
+  @VisibleForTesting
+  void tryLock(Path src, Path dst) {
+    if (!lockManager.acquire(dst.toString(), src.toString())) {
+      throw new CommitFailedException(
+          "Failed to acquire lock on file: %s with owner: %s", dst, src);
+    }
+  }
 
-    // update the best-effort version pointer
-    writeVersionHint(nextVersion);
+  void unlock(Path src, Path dst) {
+    try {
+      if (!lockManager.release(dst.toString(), src.toString())) {
+        LOG.warn("Failed to release lock on file: {} with owner: {}", dst, src);
+      }
+    } catch (Exception ignored) {
+      // do nothing.
+    }
+  }
 
-    deleteRemovedMetadataFiles(base, metadata);
+  private void validate(
+      boolean supportGlobalLocking,
+      int previousVersionsMax,
+      int nextVersion,
+      FileSystem fs,
+      Path finalMetadataFile)
+      throws IOException {
+    if (!supportGlobalLocking) {
```

Comment:
For cases where distributed locks are not used, we need to do some checksumming and clean up some dirty data.